### PR TITLE
Tasks: adds Quick Add feature

### PR DESCRIPTION
**Pull request by AbleLincoln**
Adds a feature to quickly add a new task straight from the tasks list

### CODE REVIEW

File: src/Pim/PimMain.tsx [hunk 1]
```diff
@@ -34,6 +36,8 @@ interface PropsType {
   location?: Location;
   history?: History;
   theme: Theme;
+  collectionsTaskList: EteSync.CollectionInfo[];
+  onItemSave: (item: PimType, journalUid: string, originalContact?: PimType) => void;
```
Comment (tasn) [inline src/Pim/PimMain.tsx]: Could you please fix the naming here? It should be `originalTask`, or better yet, just `originalItem`.

**Comment by tasn** [comment]
Oh I also just noticed Github is complaining that this branch has conflicts. Could you please take a look at that too? I think it's because I added the licensing information to files you renamed. Actually though, this reminds me, could you please make sure to include the same licensing information on the top of your newly added files too? This is what needs to be there:

```
// SPDX-FileCopyrightText: © 2017 EteSync Authors
// SPDX-License-Identifier: AGPL-3.0-only
```

**Comment by AbleLincoln** [comment]
Like you said, I believe the conflicts are from moving files. GitHub doesn't allow me to resolve the conflicts. If you just accept the incoming changes it should be fine, I completely copied over those conflicting files.

**Comment by tasn** [comment]
I'll resolve things locally, and merge. No problem.

**Comment by tasn** [comment]
Merged, thanks!

**Comment by tasn** [comment]
There are two comments I missed (no worries, I'm fixing them now), just FYI:
1. When naming function, we prefer using `function handleChange()` over `const handleChange = () => ...`
2. I think it's better to use a form element for handling the submission rather than hooking on the keypress event.
3. It was crashing when there were no task lists.
4. onItemSave was not passed.
5. Saving an item returns.

And a few other things. It's very much apparent that you haven't tested it at all. I'm quite pissed tbh for having wasted as much time as I have on this. And I'm annoyed with myself for trusting that you have tested it before pushing it. I'm reverting all of the changes. Please fix it, as it's completely broken.
I understand you took it out to its own PR so things have changed since you first wrote it, but there's no excuse to not testing! (I'm saying that to myself too!)

Additionally, please take a look at #32, the quickadd bar should be at the top, outside of the list, not a part of it... It's quite ugly this way, and much less functional with long lists...

**Comment by AbleLincoln** [comment]
I was actually just coming back now to add some code I realized I forgot to push. There are just some props missing that need to be passed down.

**Comment by AbleLincoln** [comment]
And I can change where the quick add bar is located, I was just going off of the design for some other apps I had researched (ios notes, Todoist).

**Comment by tasn** [comment]
I edited my comment above a bit to clarify things, and tone down my frustration.

Yeah, please the position. In general, follow #32... Also, happy to answer questions if you have any. :)

**Comment by tasn** [comment]
Additionally, please let me know once you'd like me to take a look.

**Comment by AbleLincoln** [comment]
Ok, sorry for missing things. It's a bit difficult trying to cherry-pick out the relevant code I need for each merge. Also, design is a little wonky as I add in each individual feature. I promise everything looks better once it is all together.

**Comment by tasn** [comment]
Yeah, I get, no worries. And sorry for showing my frustration, though I was just fixing one issue after the other and getting more and more annoyed until I gave up. :|

Let me know when you'd like me to take another look, though please make sure you both test and push before, and in addition to addressing the above comments.

Thanks!

**Comment by tasn** [comment]
To be honest, I wouldn't return history from the `onItemSave` - it's overly specific. You should just return a Promise and if it succeeds you should go back. The history can be gotten locally with `withRouter` or can just be passed by props.

**Comment by AbleLincoln** [comment]
I think I got everything, please take a look when you get a chance

File: src/Pim/index.tsx [hunk 2]
```diff
@@ -107,7 +107,7 @@ type CollectionRoutesPropsType = RouteComponentProps<{}> & {
   componentEdit: any;
   componentView: any;
   items: {[key: string]: PimType};
-  onItemSave: (item: PimType, journalUid: string, originalContact?: PimType) => void;
+  onItemSave: (item: PimType, journalUid: string, originalItem?: PimType) => Promise<any>;
```
Comment (tasn) [inline src/Pim/index.tsx]: Should be Promise<void> not any.

File: src/Pim/index.tsx [hunk 3]
```diff
@@ -235,11 +237,11 @@ class Pim extends React.PureComponent {
     this.onItemSave = this.onItemSave.bind(this);
   }
 
-  public onItemSave(item: PimType, journalUid: string, originalEvent?: PimType) {
+  public onItemSave(item: PimType, journalUid: string, originalEvent?: PimType): Promise<History<any>> {
```
Comment (tasn) [inline src/Pim/index.tsx]: This shouldn't return `Promise<History>` but `Promise<void>`

File: src/Pim/index.tsx [hunk 4]
```diff
@@ -251,7 +253,7 @@ class Pim extends React.PureComponent {
     if (last) {
       prevUid = last.uid;
     }
-    store.dispatch<any>(fetchEntries(this.props.etesync, journal.uid, prevUid))
+    return store.dispatch<any>(fetchEntries(this.props.etesync, journal.uid, prevUid))
```
Comment (tasn) [inline src/Pim/index.tsx]: According with my comment above, this function should not early return. Just let it return like before (by getting to the end of the function). Same with the line below.

File: src/components/ContactEdit.tsx [hunk 5]
```diff
@@ -113,9 +115,10 @@ interface PropsType {
   collections: EteSync.CollectionInfo[];
   initialCollection?: string;
   item?: ContactType;
-  onSave: (contact: ContactType, journalUid: string, originalContact?: ContactType) => void;
+  onSave: (contact: ContactType, journalUid: string, originalContact?: ContactType) => Promise<any>;
```
Comment (tasn) [inline src/components/ContactEdit.tsx]: Same here with `Promise<void>` - same with all of these places, I won't comment for each.

File: src/components/ContactEdit.tsx [hunk 6]
```diff
@@ -309,7 +312,10 @@ class ContactEdit extends React.PureComponent<PropsType> {
     setProperty('title', this.state.title);
     setProperty('note', this.state.note);
 
-    this.props.onSave(contact, this.state.journalUid, this.props.item);
+    this.props.onSave(contact, this.state.journalUid, this.props.item)
```
Comment (tasn) [inline src/components/ContactEdit.tsx]: I love it. Looks much nicer! Sorry you had to deal with the mess that was there before.

File: src/Pim/index.tsx (hunk 2 above)
Comment (AbleLincoln) [reply src/Pim/index.tsx]: `Promise<void>`?

File: src/Pim/index.tsx (hunk 4 above)
Comment (AbleLincoln) [reply src/Pim/index.tsx]: But then how can we await the finishing of the asynchronous action?

Comment (tasn) [reply src/Pim/index.tsx]: Sorry, I thought the the function was async. Either make the function async and change the `return Promise.reject()` to `throw new Error('journalUid is undefined')` and await on these two async calls (and everything else will just work). Or you can manually create a promise. Though I think making the function async is much easier.

Comment (tasn) [reply src/Pim/index.tsx]: To be honest though, the returns you did are fine too if changing the return type to `Promise<void>` won't complain about the types.

File: src/Pim/index.tsx (hunk 2 above)
Comment (tasn) [reply src/Pim/index.tsx]: Sorry, that's what I meant, yeah. I forgot to wrap it in code tags so the `<void>` part was stripped. It seems.

Comment (AbleLincoln) [reply src/Pim/index.tsx]: Ah yes, my favorite HTML tag: `<void>`

File: src/Pim/index.tsx (hunk 4 above)
Comment (AbleLincoln) [reply src/Pim/index.tsx]: Yeah TypeScript doesn't seem to have a problem with the return type. `store.dispatch` is weird, it returns whatever you pass in, which in this case is of the type `Action<Promise<EteSync.Entry[]>>`. Which I guess resolves to a Promise? I'm not sure, it's a bit confusing, but it works haha.

Honestly I am ok with either way (returning promises or marking it async), it seems more like a style choice at this point. Which would you prefer?

Comment (tasn) [reply src/Pim/index.tsx]: It's typing issue which I haven't managed to solve. The return type is indeed as you said, but the actual return type is `Promise<Action<EteSync.Entry[]>>`. That's why there's an `<any>` there to workaround this issue. :|
If this works then no need to change it, it's all the same.

File: src/Pim/index.tsx (hunk 2 above)
Comment (tasn) [reply src/Pim/index.tsx]: hehe. :laughing:

**Comment by tasn** [comment]
Let me know when you'd like me to take another look at the code. Thanks!

**Comment by AbleLincoln** [comment]
It's good to go

**Comment by tasn** [comment]
I think you forgot to push, all of the promises are still not `Promise<void>`.

**Comment by AbleLincoln** [comment]
Where? Did I miss one?

**Comment by tasn** [comment]
Nah, all good. For some reason it didn't show me you pushed anything when I replied, I was still seeing the old diff.

**Comment by tasn** [comment]
You added a few weird files, such as: `src/components/Tasks/TaskEdit.tsx~refs/remotes/upstream/master`.
Not sure where they came from, but please remove them.

**Comment by AbleLincoln** [comment]
My bad, have no idea how that happened. Is this ok now?

**Comment by tasn** [comment]
In, good job!
I had to fix one thing (you forgot to change one `onItemSave => void` to return `Promise<void>`).
I also squashed most of the commits together as we don't really want to litter the history with changes that are going back and forth between ideas. It's fine on the other hand to have multiple commits for different parts or evolution of a complex feature (step 1, 2 and etc.).

Unfortunately github doesn't detect merges when they happen from the command line, so I have to close this manually unfortunately, though the commits are in!

**Pull request closed without merging**